### PR TITLE
Monitoring VM Status Transition Times

### PR DESCRIPTION
**Pull request by marceloamaral**
**What this PR does / why we need it**:
This PR introduces a new VM status field that tracks the status transition timestamps. It also adds a new histogram metric to help track how long it takes for status transitions to occur.

This is similar the VMI phase transition metric introduced in the PR https://github.com/kubevirt/kubevirt/pull/5766, but for the VM object.

The new histogram metric `kubevirt_vm_status_transition_time_from_creation_seconds` tracks how long since creation it took to transition to the current status a VM is in. To be esier to follow, we use the PrintableStatus. This would be useful if you wanted to track how long it takes for VM to go from Creation to Running (completed).

Using variations of that query, we can create graphs that indicate how long it takes for VMs to reach running, and which state the VM spends the most time in.
image

Only analyzing the VMI phase transition time does not show the latency of the VM object. As we can see in the figure below, the even though the VMI entered in the Ready state, the VM took much longer to also enter in the Ready state.
image

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
Addition of kubevirt_vm_status_transition_time_from_creation_seconds metric to monitor how long it takes to transition a VM object to a specific state from creation time.
```

### CODE REVIEW

**Comment by iholder-redhat** [comment]
Thanks @marceloamaral!

1. What about cleaning `StatusTransitionTimestamps`? With VMIs the amount of transitions is pretty small, but with VMs this is not the case. A VM can have a very long life-cycle with many different VMIs.
2. I think your commits are in reversed order

File: pkg/monitoring/perfscale/vm-status-transitions.go [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+	oldTime = newVM.DeletionTimestamp.DeepCopy()
+	if oldTime == nil {
+		oldTime = newVM.CreationTimestamp.DeepCopy()
+	}
+	for _, transitionTimestamp := range newVM.Status.StatusTransitionTimestamps {
+		if transitionTimestamp.Status == newVM.Status.PrintableStatus {
+			newTime = transitionTimestamp.StatusTransitionTimestamp.DeepCopy()
+		} else if newTime != nil {
+			break
+		}
```
Comment (davidvossel) [inline pkg/monitoring/perfscale/vm-status-transitions.go:48]: Hey @marceloamaral

sorry i'm delayed in giving feedback here. 

I'm unsure if we should be anchoring a metric to thie vm.Status.PrintableStatus value. My concern is that `PrintableStatus` is only meant to serve as a best effort representation of the state a VM is that is used to provide user feedback via CLI and UI.

PrintableStatus is not a programmatic interface that is guaranteed to behave in any sort of consistent way throughout kubevirt releases. It's possible for this field to get new values added, removed, modified, etc... We maintain the freedom to modify how this field is calculated as well in ways that could impact the transition times in the future.

In contrast to PrintableStatus, the vmi.Status.Phase field is meant to be a programmatic interface that behaves consistently. We made the decision to limit the phases to only the values that exist today, and to express any additional information outside of the phases using conditions. This is why having a metric associated with vmi.Status.Phase is likely consistent throughout future kubevirt releases.

The other kind of weird thing about PrintableStatus is that unlike vmi.Status.Phase, we can re-encounter transitioning to the same PrintableStatus many times throughout a VM's lifecycle. Even during the start flow, if an transient error is encountered, it's possible for the PrintableStatus field to transition to the same value multiple times before hitting "Running"

What metric is most important to you here? Perhaps there's another way to track it?

File: staging/src/kubevirt.io/api/core/v1/types.go [hunk 2]
```diff
@@ -1372,6 +1372,14 @@ const (
...
 )
 
+// VirtualMachineStatusTransitionTimestamp gives a timestamp in relation to when a status is set on a vm
+type VirtualMachineStatusTransitionTimestamp struct {
+	//  Hold the state information of the VirtualMachine.
+	Status VirtualMachinePrintableStatus `json:"status,omitempty"`
+	// StatusTransitionTimestamp is the timestamp of when the status change occurred
+	StatusTransitionTimestamp metav1.Time `json:"statusTransitionTimestamp,omitempty"`
+}
+
```
Comment (davidvossel) [inline staging/src/kubevirt.io/api/core/v1/types.go:1382]: For printableStatus, the same status field can be encountered many times throughout the VMs lifecycle.

**Pull request closed without merging**